### PR TITLE
Add auto shard aware port

### LINKED ISSUE #255: Automatically use the shard-aware port when available
As mentioned in https://github.com/scylladb/scylla/blob/f186de909d461003e83ce24f256c2fb6884b6004/docs/design-notes/protocol-extensions.md#intranode-sharding,  Scylla responds to CQL's `OPTIONS` request with a `SCYLLA_SHARD_AWARE_PORT` entry, which describes the port used for the new shard awareness mechanism.

Instead of making the users set up and type the `19042` port manually, it should be handled automatically as follows:
- user connects to the regular CQL port - 9042
- user sends `OPTIONS` and receives `SCYLLA_SHARD_AWARE_PORT` and `SCYLLA_SHARD_AWARE_PORT_SSL` for secure connections
- user reestablishes the connection using the acquired shard-aware port

### PR DESCRIPTION
**Pull request by psarna**
This patch makes sessions automatically probe for shard-aware ports
via the CQL OPTIONS request, and if found, tries to use it.
With such a mechanism, users no longer need to remember to manually
set up port 19042 to enjoy shard awareness.

The new algorithm is as follows:
 - connect to one of the nodes just to send an `OPTIONS` request and retrieve the shard-aware port
   - if found, use the shard aware port to connect to the cluster
     - if successful, continue enjoying the driver
     - if failed, fall back to the original port (a custom port can fail e.g. due to firewall rules)
   - if not found, use the original port

Currently marked as draft, because the change was only tested manually in a limited scope. Still, it's ready for preliminary review.

Fixes #255

## Pre-review checklist

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [ ] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [x] I added appropriate `Fixes:` annotations to PR description.

### CODE REVIEW

**Comment by psarna** [comment]
note: the pull request was created with programmer's ergonomics in mind. In particular, if the client already passed the shard-aware port, it will still be probed anyway. It creates a minimal overhead, but I consider it negligible, because creating new sessions is not in any way a most common operation during applications' execution time.

**Comment by psarna** [comment]
v2:
 - changed a stray `println!` used for debugging to `info!`
 - tested manually with Scylla, Cassandra and checking the logs:

Tested with `RUST_LOG=scylla=trace cargo run --example logging`

Cassandra:

```
Jun 25 09:28:01.085  WARN scylla::transport::session: Unable to establish connections at detected shard-aware port, falling back to default ports
Jun 25 09:28:01.086 DEBUG scylla::transport::cluster: Requesting topology refresh
Jun 25 09:28:01.139  WARN scylla::transport::session: Raw USE KEYSPACE queries are experimental, please use session::use_keyspace instead
```

Scylla:
```
Jun 25 09:29:13.493  INFO scylla::transport::session: Shard-aware port detected: 19042
Jun 25 09:29:13.493 DEBUG scylla::transport::cluster: Requesting topology refresh
Jun 25 09:29:13.502  WARN scylla::transport::session: Raw USE KEYSPACE queries are experimental, please use session::use_keyspace instead

```

For Scylla I also verified that the connection is shard-aware via cqlsh:
```bash
cqlsh 127.0.0.3 <<< "select address, driver_name, shard_id from system.clients where driver_name = 'scylla-rust-driver' allow filtering;"

 address   | driver_name        | shard_id
-----------+--------------------+----------
 127.0.0.1 | scylla-rust-driver |        2
 127.0.0.1 | scylla-rust-driver |        3
 127.0.0.1 | scylla-rust-driver |        0
 127.0.0.1 | scylla-rust-driver |        1
```

File: scylla/src/transport/session.rs [hunk 1]
```diff
@@ -283,8 +283,40 @@ impl Session {
...
+            Cluster::new(&shard_aware_addresses, config.get_connection_config())
+                .await
+                .ok()
+        } else {
+            None
+        };
+        let cluster = match shard_aware_cluster {
+            Some(clust) => clust,
+            None => {
+                warn!("Unable to establish connections at detected shard-aware port, falling back to default ports");
```
Comment (piodul) [inline scylla/src/transport/session.rs:316]: Here, maybe we could print what error happened while attempting to connect to the shard-aware port?

**Review by piodul — COMMENTED** [review]
Looks good, but needs a rebase first - there are conflicts in `scylla/src/transport/session.rs`.

File: scylla/src/transport/session.rs (hunk 1 above)
Comment (psarna) [reply scylla/src/transport/session.rs]: Good idea, done. I collapsed these two pattern matching statements into one

**Comment by psarna** [comment]
v3:
 - preserved the error message from not being able to connect to the shard-aware port
 - rebased

**Review by piodul — APPROVED** [review]

**Pull request merged**